biz.aQute.repository/src/aQute/bnd/repository/p2/provider/P2Indexer.java

```diff
@@ -109,24 +114,44 @@ File get(String bsn, Version version, Map<String, String> properties, DownloadLi
 			return null;
 
 		URI url = contentCapability.url();
+		boolean packed = url.getPath()
+			.endsWith(Unpack200.PACKED_SUFFIX);
 
-		final File source = client.getCacheFileFor(url);
 		final File link = new File(location, bsn + "-" + version + ".jar");
 
-		IO.createSymbolicLinkOrCopy(link, source);
-
 		Promise<File> go = client.build()
 			.useCache(MAX_STALE)
 			.async(url.toURL())
-			.map(file -> link);
+			.then(success -> success.map(file -> handleDownloadResult(packed, link, success.getValue())));
 
-		if (listeners.length == 0)
+
+		if (listeners.length == 0) {
 			return go.getValue();
+		}
 
 		new DownloadListenerPromise(reporter, name + ": get " + bsn + ";" + version + " " + url, go, listeners);
+
 		return link;
 	}
 
+	private File handleDownloadResult(boolean packed, File link, File source)
+		throws Exception {
+
+		File unpacked = new File(source.getParentFile(), source.getName() + ".unpacked");
+		if (packed && !unpacked.exists()) {
```

Comment:
I think the exists() test is a problem. It means that if the remote file is changed we never unpack the changed file it if was previously unpacked. That seems like a bad caching problem.

Also, we do we keep both files (original and original.unpacked). Perhaps we should unpack in place? Then we can know the link value before and avoid having to create it later here.